### PR TITLE
Add live progress for runs

### PR DESCRIPTION
**Pull request by alexandrudanpop**
Fixes OPS-1123

## Additional Notes
Update requests are sent to client through websocket. Only the runId is sent through the WS. The react client will check if the user views the runId from the WS and it will refetch it. If there are too many requests from WS to the react-app while requests to get runs are pending previous requests will be cancelled

## Testing Checklist

Check all that apply:

- [x] I tested the feature thoroughly, including edge cases

- [x] I verified all affected areas still work as expected

- [x] Automated tests were added/updated if necessary

- [x] Changes are backwards compatible with any existing data, otherwise a migration script is provided

## Visual Changes (if applicable)

Deployed on test.
https://www.loom.com/share/30eef78de0e94c62be5e5fc59ee52c13

### CODE REVIEW

File: packages/server/api/src/app/workers/engine-controller.ts [hunk 1]
```diff
@@ -135,15 +147,28 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
...
+      !isNil(executionStateContentLength) && executionStateContentLength > 0;
+    if (!updateLogs) {
+      if (progressUpdateType === ProgressUpdateType.TEST_FLOW) {
+        app.io
+          .to(request.principal.projectId)
+          .emit(WebsocketClientEvent.TEST_FLOW_RUN_PROGRESS, runId);
+      } else {
+        app.io
+          .to(request.principal.projectId)
+          .emit(WebsocketClientEvent.FLOW_RUN_PROGRESS, runId);
```
Comment (alexandrudanpop) [inline packages/server/api/src/app/workers/engine-controller.ts]: I wonder if we need both `TEST_FLOW_RUN_PROGRESS` and `FLOW_RUN_PROGRESS`, I would keep only `FLOW_RUN_PROGRESS` unless we want to have some telemetry based on them?

Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts]: I received info from Marco, that scheduled runs (production runs don't need live update). This will make it better from a technical perspective as we don't send too many websocket requests in case there are many concurrent scheduled runs.

File: jest.env.js [hunk 2]
```diff
@@ -1,2 +1,2 @@
-
 process.env.OPS_QUEUE_MODE = 'MEMORY';
+process.env.OPS_ENGINE_URL = 'http://localhost:3005';
```
Comment (cezudas) [inline jest.env.js:2]: Is it needed?
Does `packages/server/api/test/unit/engine/engine-controller.test.ts` or `packages/engine/test/services/progress.service.test.ts` fail without it?

File: packages/react-ui/src/app/features/builder/hooks/use-run-progress.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import { useSocket } from '@/app/common/providers/socket-provider';
+import { flowRunsApi } from '@/app/features/flow-runs/lib/flow-runs-api';
+import { FlowRun, FlowVersion, WebsocketClientEvent } from '@openops/shared';
+
+interface UseRunProgressParams {
+  run: FlowRun | null;
+  setRun: (run: FlowRun, flowVersion: FlowVersion) => void;
+  flowVersion: FlowVersion | null;
+}
```
Comment (rmn-snpk) [inline packages/react-ui/src/app/features/builder/hooks/use-run-progress.ts]: [nitpick] we use `type` in most of the files

File: jest.env.js (hunk 2 above)
Comment (alexandrudanpop) [reply jest.env.js:2]: yep, we need it for `engine-controller` tests

File: packages/server/api/src/app/workers/engine-controller.ts [hunk 4]
```diff
@@ -135,15 +136,14 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
       tags: runDetails.tags ?? [],
     });
 
-    if (populatedRun.status === FlowRunStatus.RUNNING) {
-      return;
-    }
```
Comment (MarceloRGonc) [inline packages/server/api/src/app/workers/engine-controller.ts:140]: Don't remove this. Move this code to after the web-socket event, but before marking the job as complete.

File: packages/server/api/src/app/workers/engine-controller.ts [hunk 5]
```diff
@@ -135,15 +136,14 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
...
-    if (runDetails.status === FlowRunStatus.PAUSED) {
+    if (
+      runDetails.status === FlowRunStatus.PAUSED &&
+      !isNil(runDetails.pauseMetadata)
+    ) {
       await flowRunService.pause({
         flowRunId: runId,
         pauseMetadata: {
-          ...runDetails.pauseMetadata!,
+          ...runDetails.pauseMetadata,
```
Comment (MarceloRGonc) [inline packages/server/api/src/app/workers/engine-controller.ts]: ...(runDetails.pauseMetadata ?? {}),

File: packages/server/api/src/app/workers/engine-controller.ts [hunk 6]
```diff
@@ -135,15 +136,14 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
...
     });
 
-    if (populatedRun.status === FlowRunStatus.RUNNING) {
-      return;
-    }
-
-    if (runDetails.status === FlowRunStatus.PAUSED) {
+    if (
+      runDetails.status === FlowRunStatus.PAUSED &&
+      !isNil(runDetails.pauseMetadata)
```
Comment (MarceloRGonc) [inline packages/server/api/src/app/workers/engine-controller.ts]:
```suggestion
      runDetails.status === FlowRunStatus.PAUSED
```

Comment (rita-gorokhod) [inline packages/server/api/src/app/workers/engine-controller.ts]: how can it be null and paused at the same time? if it's paused, we have to have the pauseMetadata 🤔

File: packages/server/api/src/app/workers/engine-controller.ts [hunk 7]
```diff
@@ -152,9 +152,10 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
         },
       });
     }
+
     app.io
-      .to(populatedRun.projectId)
-      .emit(WebsocketClientEvent.TEST_FLOW_RUN_PROGRESS, populatedRun);
+      .to(request.principal.projectId)
+      .emit(WebsocketClientEvent.FLOW_RUN_PROGRESS, runId);
```
Comment (rita-gorokhod) [inline packages/server/api/src/app/workers/engine-controller.ts:155]: I'm curious why we're not passing the populatedRun as before, what was the problem?

File: packages/react-ui/src/app/features/flows/lib/flows-api.ts [hunk 8]
```diff
@@ -172,7 +142,7 @@ function getInitialRun(
   socket: Socket,
   flowVersionId: string,
 ): Promise<FlowRun> {
-  return new Promise<FlowRun>((resolve, reject) => {
```
Comment (rita-gorokhod) [inline packages/react-ui/src/app/features/flows/lib/flows-api.ts:175]: can you explain why this was removed?

File: packages/engine/src/lib/handler/flow-executor.ts [hunk 9]
```diff
@@ -77,20 +77,14 @@ export const flowExecutor = {
         break;
       }
 
-      const isNotNested = flowExecutionContext.currentPath.path.length === 0;
-      const sendContinuousProgress =
-        isNotNested &&
-        constants.progressUpdateType === ProgressUpdateType.TEST_FLOW;
```
Comment (alexandrudanpop) [inline packages/engine/src/lib/handler/flow-executor.ts]: we don't need those filters if we want continuous progress when executing the flow

File: packages/engine/src/lib/services/progress.service.ts [hunk 10]
```diff
@@ -1,52 +1,70 @@
...
-
-      await fetch(url.toString(), {
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/json',
-          Authorization: `Bearer ${engineConstants.engineToken}`,
-        },
-        body: JSON.stringify(request),
-      });
+      await sendUpdateRunRequest(params);
```
Comment (alexandrudanpop) [inline packages/engine/src/lib/services/progress.service.ts:16]: Just extracted sendUpdateRunRequest to a separate function, and changed the implementation to do request deduplication like in AP

File: packages/server/api/src/app/workers/engine-controller.ts (hunk 7 above)
Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts:155]: I don't think it's a good practice to send large payloads to the clients through websockets. For example:
React app user is viewing Run1, but engine is sending updates because it's executing Run2, now those payloads for Run2 are send to the app. With this approach it just receives a runId instead of a full run payload.

File: packages/react-ui/src/app/features/flows/lib/flows-api.ts (hunk 8 above)
Comment (alexandrudanpop) [reply packages/react-ui/src/app/features/flows/lib/flows-api.ts:175]: This was previously supposed to track the flow run progress, now we the hook from `use-run-progress` to track the progress and this just initiates the run.

File: packages/react-ui/src/app/features/builder/hooks/use-run-progress.ts (hunk 3 above)
Comment (alexandrudanpop) [reply packages/react-ui/src/app/features/builder/hooks/use-run-progress.ts]: fixed

File: packages/server/api/src/app/workers/engine-controller.ts (hunk 4 above)
Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts:140]: fixed

File: packages/server/api/src/app/workers/engine-controller.ts (hunk 5 above)
Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts]: done

File: packages/server/api/src/app/workers/engine-controller.ts (hunk 6 above)
Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts]: done

Comment (alexandrudanpop) [reply packages/server/api/src/app/workers/engine-controller.ts]: please re-check

File: packages/engine/src/lib/services/progress.service.ts [hunk 11]
```diff
@@ -1,52 +1,70 @@
...
+    workerHandlerId: engineConstants.serverHandlerId ?? null,
+    runDetails: await flowExecutorContext.toResponse(),
+    progressUpdateType: engineConstants.progressUpdateType,
+  };
+
+  // Request deduplication using hash comparison
+  const requestHash = crypto
+    .createHash('sha256')
+    .update(JSON.stringify(request))
+    .digest('hex');
```
Comment (MarceloRGonc) [inline packages/engine/src/lib/services/progress.service.ts]: You can use the hashObject method. Also this is not working because the duration time changes.
```suggestion
  const requestHash = hashUtils.hashObject(request, (key, value) => {
    if (key === 'duration') return undefined;
    return value;
  });
```

File: packages/engine/src/lib/handler/flow-executor.ts [hunk 12]
```diff
@@ -77,20 +77,14 @@ export const flowExecutor = {
...
-        progressService
-          .sendUpdate({
-            engineConstants: constants,
-            flowExecutorContext: flowExecutionContext,
-          })
-          .catch((error) => {
-            logger.error('Error sending progress update', error);
-          });
-      }
+      progressService
```
Comment (rita-gorokhod) [inline packages/engine/src/lib/handler/flow-executor.ts]: I see that we're doing this after "if (flowExecutionContext.verdict !== ExecutionVerdict.RUNNING) {", so we won't get to it if it was failed or paused, no?

File: packages/engine/src/lib/services/progress.service.ts (hunk 11 above)
Comment (alexandrudanpop) [reply packages/engine/src/lib/services/progress.service.ts]: fixed, thanks

File: packages/engine/src/lib/handler/flow-executor.ts (hunk 12 above)
Comment (alexandrudanpop) [reply packages/engine/src/lib/handler/flow-executor.ts]: adjusted, I'll deploy again to test

**Review by rSnapkoOpenOps — APPROVED** [review]

File: packages/engine/src/lib/services/progress.service.ts [hunk 13]
```diff
@@ -1,52 +1,69 @@
...
+    workerHandlerId: engineConstants.serverHandlerId ?? null,
+    runDetails: await flowExecutorContext.toResponse(),
+    progressUpdateType: engineConstants.progressUpdateType,
+  };
+
+  // Request deduplication using hash comparison
+  const requestHash = hashUtils.hashObject(request, (key, value) => {
+    if (key === 'duration') return undefined;
+    return value;
+  });
```
Comment (MarceloRGonc) [inline packages/engine/src/lib/services/progress.service.ts:49]: Because we are calling executeFromAction recursively, the event we are sending may have already been sent. We are hashing the request to validate if we have already sent the event and reduce the number of requests.

Example: We have a branch with a code block inside.

executeFromAction (Branch) -> executeFromAction (Code)

When the code block finishes, it calls the sendUpdateRunRequest and returns the result. Then the Branch finishes and calls the sendUpdateRunRequest with the same result. The only difference is the duration.
